util/ntp/ntp.go

```diff
@@ -0,0 +1,134 @@
+package ntp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/beevik/ntp"
+	"github.com/pactus-project/pactus/util/logger"
+)
+
+const (
+	maxClockOffset = time.Duration(math.MinInt64)
+)
+
+type Checker struct {
+	lock   sync.RWMutex
+	ctx    context.Context
+	cancel func()
+
+	threshold  time.Duration
+	offset     time.Duration
+	interval   time.Duration
+	logger     *logger.SubLogger
+	retryTimes int
+}
+
+func NewNtpChecker(interval time.Duration) *Checker {
+	ctxWithCancel, cancel := context.WithCancel(context.Background())
+	server := &Checker{
+		ctx:        ctxWithCancel,
+		cancel:     cancel,
+		interval:   interval,
+		threshold:  1 * time.Second,
+		retryTimes: 3,
+	}
+
+	server.logger = logger.NewSubLogger("_ntp", server)
+
+	return server
+}
+
+func (c *Checker) clockOffset() time.Duration {
+	clockOffset := maxClockOffset
+
+	for i := 0; i < c.retryTimes; i++ {
+		server := "pool.ntp.org"
+		response, err := ntp.Query(server)
+		if err != nil {
+			c.logger.Debug(
+				fmt.Sprintf("ntp error(retry %v/%v times)", i+1, c.retryTimes),
+				"server", server, "error", err,
+			)
+
+			continue
+		}
+
+		if err := response.Validate(); err != nil {
+			c.logger.Debug(
+				fmt.Sprintf("ntp error(retry %v/%v times)", i+1, c.retryTimes),
+				"server", server, "error", err,
+			)
+
+			continue
+		}
+
+		clockOffset = response.ClockOffset
+
+		break
+	}
+
+	return clockOffset
+}
+
+func (c *Checker) checkClockOffset() {
+	ticker := time.NewTicker(c.interval)
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+
+		case <-ticker.C:
+			c.lock.Lock()
```

Comment:
This has flaw.
Imagine `c.clockOffset` took time to process. and we locked till it finishes.
Therefore `GetClockOffset()` will be blocked.